Support Airtable API

Xheldon opened the pull request:
issue https://github.com/obsidianmd/obsidian-importer/issues/455

This PR’s initial submission only implements the basic functionality; formula, Rollup, Lookup, and other related properties have not been properly handled yet. These features will be fixed in later code commits.

If anyone is willing to try this PR, please be sure to back up your Vault.

Some of the code was generated by AI. I have reviewed and tested it myself, but there are parts that need refactoring. I will continue to optimize the code in subsequent commits and will request the author’s review when the timing feels right.

Xheldon committed:
Airtable API: Basic UI

Refer to the Notion API workflow to implement basic interfaces such as a directory tree (Base/Table) selection.

Xheldon committed:
Airtable API: Add TemplateConfigurator

Refer to the CSV import workflow and display the field template configuration interface.

Xheldon committed:
Airtable API: Some adjust

* Add incremental import toggle
* Add Base link field setting
* Cache certain API requests
* Make the Note location setting in CSV templates configurable (since we’ve already created folders by Base/Table)

Xheldon committed:
Airtable API: View should not be importable on its own

Xheldon committed:
Airtable API: Handle some cases

Xheldon committed:
Airtable API: Optimize import speed

Xheldon committed:
Airtable API: Set property to types.json

Xheldon committed:
Airtable API: Refactor the import process

Considering that Airtable’s API offers only a few request types (listing bases, fetching a table schema, retrieving view records), I split the import process into two parts: process the data first, then write it to files.

The import process is now divided into two stages:
Stage 1: Retrieve the records of all tables in all bases, along with the corresponding view data, and build the correct content, references, formulas, etc.
Stage 2: Write the files, including all view information into a .base file (checking for existing files before writing).

Note that any error in Stage 1 will abort the import.

Xheldon committed:
Airtable API: Handle Formula(WIP)

The complete set of formulas supported by Airtable and examples can be found at: https://airtable.com/appCHsByjT6czDeJu/shrHgArBZ61sNOtKl/tblkwFqAgvq4M5tp6/viwAD5P14cSxKntDm?blocks=hide

Xheldon committed:
Airtable API: Handle Text Formula(WIP-2)

Properly handle attribute conversion
Properly handle formula conversion (WIP)

Xheldon committed:
Airtable API: Check text formula one by one

* The field order now matches the order of fields in Airtable
* Check the Text function conversion logic one by one to ensure it is correct.

Xheldon committed:
Airtable API: Check logic formula one by one

* Correctly handle rename logic
* Handle illegal string filename logic
* Correctly handle error flow
* Proper handling `=` the differences between Airtable and Obsidian

Xheldon committed:
Airtable API: Handle boundary cases in logic functions

Xheldon committed:
Airtable API: Check number formula one by one

Xheldon committed:
Airtable API: Check date formula one by one

Xheldon committed:
Airtable API: Check the remaining formulas one by one

Ready to test the all formula!

Xheldon committed:
Airtable API: Handle Linked Record related logic

Xheldon committed:
Airtable API: Lint error fix

Xheldon committed:
Airtable API: Attachment uses user settings (excluding YAML properties)

Xheldon committed:
Airtable API: Optimize the code

* remove legacy parameters
* Use the same directory tree style class names as the Notion API

Xheldon committed:
Airtable API: Remove unnecessary branches and logic

Xheldon committed:
Airtable API: Handle more details

* Remove unused functions
* Remove unused parameters
* Extract function parameters of five or more into types
* Special handling for Currency and Rating types
* Handle type issues

Xheldon committed:
Airtable API: Fix YAML property formatting issue

Xheldon committed:
Airtable API: Refactor Base import process

Previously: After retrieving all complete information, assemble the data in memory and then write it to a file.
Now: Assemble the data by Base and then write it, to reduce memory usage.
Note: Because in Airtable, RecordId and FieldId are only guaranteed to be unique within the same Base, we prepend a BaseId when using them.

Xheldon committed:
Airtable API: Some logical adjustments

Remove the Note Title template setting logic, since we’ll have multiple tables with potentially different fields; now we uniformly use the Table’s primary field as the filename.
Do not handle Currency and Rating specially; treat both as numeric types.

Xheldon committed:
Airtable API: Process the body’s attachments

tgrosinger reviewed (commented):
Finished reading through the main logic, but I still need to review the helper/converter files.

Most of my comments are minor code style. I have not tested it yet, but the logic seems solid.

Looking really great so far. Thank you @Xheldon!

tgrosinger commented on src/formats/airtable-api/attachment-helpers.ts:
```diff
@@ -0,0 +1,224 @@
+		const pathWithoutExt = targetPath.replace(/\.[^/.]+$/, '');
+		
+		return {
+			path: pathWithoutExt,
```
This doesn't feel like the right data to be returning because there are two instances below where the extension is extracted from the filename and then appended to the end of this pathWithoutExt.

Maybe just return the full path? Then the users of this data can call the various filepath utilities to get the exact format required.

Xheldon committed:
Airtable API: Handle special characters in ViewName

Xheldon committed:
Airtable API: Fixed the issue where records could not be found because the index was not updated after renaming

Xheldon committed:
Airtable API: Remove unused code

Xheldon committed:
Airtable API: Process attachment function signature

tgrosinger commented on src/formats/airtable-api/formula-converter.ts:
```diff
@@ -0,0 +1,786 @@
+	let inString = false;
+	let stringChar = '';
+	
+	for (let i = openPos + 1; i < str.length; i++) {
```
Not strictly necessary, but if you wanted you might be able to abstract some of this functionality into a helper function which walks along a string, accumulating characters and performing an action unless in quotes.

Xheldon committed:
Airtable API: Use public functions

Xheldon commented on src/formats/airtable-api.ts:
```diff
@@ -0,0 +1,2124 @@
+				.setButtonText('Select all')
+				.onClick(() => {
+					this.toggleSelectButton = toggleButtonRef;
+					this.handleToggleSelectClick();
```
`this.handleToggleSelectClick` is indeed used only once and should be placed in the callback of this `onClick`, but the assignment of `this.toggleSelectButton` cannot be removed because that value is used in the `Load` button callback (`renderTree/loadTree`). I need to access that value across functions (by default the button is hidden, and it only appears after clicking the `Load` button).

The reason for adding a `toggleButtonRef` is that I encountered a case with the Notion API earlier and asked about it on Discord; the screenshot is shown below：

<img width="1862" height="1124" alt="CleanShot_2025_12_10_at_12_04_56@2x" src="https://github.com/user-attachments/assets/0ab6dec5-a718-477c-9317-8584e106a759" />

So my solution in the Notion API was also aimed at correctly setting `this.toggleSelectButton`, using a local variable within the closure; to access the component across functions, I used a reference to `this.toggleSelectButton`.

I wonder if my explanation is clear to you?

Xheldon commented on src/formats/airtable-api.ts:
```diff
@@ -0,0 +1,2124 @@
+
+		// Create the change list container
+		this.treeContainer = publishSection.createDiv('publish-change-list');
+		this.treeContainer.style.maxHeight = '200px';
```
Yes, this is a lazy move... I directly copied the changes Kepano made to my code in the Notion API. The proper way would be to put it in `style.css`. I’ll handle this part of the Notion API styles together:

<img width="1538" height="480" alt="CleanShot_2025_12_10_at_12_13_50@2x" src="https://github.com/user-attachments/assets/45a466e8-6feb-43c8-b0cf-398d73b5173b" />

Xheldon committed:
Airtable API: Extract style

* Extract style
* Inline functions that are used only once

Xheldon commented on src/formats/airtable-api.ts:
```diff
@@ -0,0 +1,2124 @@
+		// View property name
+		new Setting(this.modal.contentEl)
+			.setName('View property name')
+			.setDesc('Property name to track which views a record belongs to. Each record will have a list of view names it appears in.')
```
The copy in this spot can finally be reviewed by @kepano for the code—--just like last time.

Xheldon commented on src/formats/airtable-api.ts:
```diff
@@ -0,0 +1,2148 @@
+	/**
+	 * Render tree UI
+	 */
+	private renderTree(): void {
```
That's a good idea; I can start working on it. `id`, `name`, `children`, and other tree nodes are generic. However, do you think the expand/collapse logic should also be extracted as generic logic? 
I'm fairly familiar with the Notion API and Airtable and can handle it, but OneNote might require a bit more effort since I'm not very familiar with the OneNote node structure.

Xheldon committed:
Airtable API: Narrowing type

Xheldon commented on src/formats/airtable-api.ts:
```diff
@@ -0,0 +1,2124 @@
+		this.loadButton.setButtonText('Loading...');
+
+		try {
+			const tempCtx = {
```
Yes, actually the ctx here only uses the `status` function; I implemented a `StatusReporter` type specifically for this.

Xheldon commented on src/formats/airtable-api.ts:
```diff
@@ -0,0 +1,2148 @@
+
+				// Toggle CSS classes and visibility
+				if (node.collapsed) {
+					collapseIcon.addClass('is-collapsed');
```
Alright, modify it together with the Notion API.

Xheldon committed:
Airtable API: Simplify the code

* Remove redundant variables
* Use toggleClass instead of add/removeClass

Xheldon committed:
Airtable API: Simplify the code2

Xheldon committed:
Airtable API: Simplify the code3

Xheldon committed:
Airtable API: Simplify the code4

Remove redundant internal functions

Xheldon committed:
Airtable API: Simplify the code5

Merge similar functions

Xheldon commented on src/formats/airtable-api.ts:
```diff
@@ -0,0 +1,2148 @@
+	/**
+	 * Toggle node selection
+	 */
+	private toggleNodeSelection(node: AirtableTreeNode, selected: boolean): void {
```
I merged `toggleNodeSelection` and `setNodeChildrenSelection`

Xheldon replied:
Done: [9d36378](https://github.com/obsidianmd/obsidian-importer/pull/461/commits/9d36378902710cfc30e161a65e555ad889404e79)

Xheldon commented on src/formats/airtable-api.ts:
```diff
@@ -0,0 +1,2148 @@
+	 */
+	private selectAllNodes(selected: boolean): void {
+		const processNode = (node: AirtableTreeNode) => {
+			if (!node.disabled) {
```
Here I kept the same logic as the Notion API: once a parent is selected, its child is automatically selected and cannot be deselected. I recommend leaving this logic unchanged, because Airtable has a Linked Records column type (similar to Notion’s Relation), and if you don’t import every table in a base, the table containing a referenced column might not be found.

Xheldon commented on src/formats/airtable-api.ts:
```diff
@@ -0,0 +1,2148 @@
+			}
+			
+			// Check if this field was converted to a formula
+			if (formulas.has(field.name)) {
```
OK, I'll rewrite it concisely using the ternary operator.

Xheldon replied:
Done: [8770b9f](https://github.com/obsidianmd/obsidian-importer/pull/461/commits/8770b9f8b33e800aca6e5225fb0d99240553af26)

Xheldon commented on src/formats/airtable-api.ts:
```diff
@@ -0,0 +1,2148 @@
+	/**
+	 * Create a single .base file for the table with multiple views
+	 */
+	private async createViewBaseFiles(ctx: BaseFileContext): Promise<void> {
```
Yes, you're right, I changed it to `createBaseFile`.

Xheldon commented on src/formats/airtable-api.ts:
```diff
@@ -0,0 +1,2148 @@
+
+				// Convert field value with direct linked record resolution
+				// Use globalFieldIdToNameMap for lookup/rollup fields that reference other tables
+				let convertedValue = convertFieldValue({
```
Yes, I added a cache and will use it directly from now on.

Xheldon replied:
Changed as above

Xheldon committed:
Airtable API: Simplify the code5

* Reduce the number of loop calls
* Rename function
* Add caching, optimize call logic

Xheldon commented on src/formats/airtable-api.ts:
```diff
@@ -0,0 +1,2148 @@
+		}
+
+		// Apply body template
+		const bodyContent = this.templateConfig
```
Exactly right, I'll tweak this part of the logic later; we should check whether `this.templateConfig.bodyTemplate` exists beforehand to avoid executing unnecessary code.

Xheldon committed:
Airtable API: Optimize execution logic and return early.

Xheldon committed:
Airtable API: Optimize attachment path formatting logic

Xheldon committed:
Airtable API: Remove duplicate formula usability detection

Xheldon committed:
Airtable API: Extract the regex as a constant

Xheldon commented on src/formats/airtable-api.ts:
```diff
@@ -0,0 +1,2124 @@
+			.setName('Select tables to import')
+			.setDesc('Click "Load" to browse your Airtable bases and tables.');
+
+		let toggleButtonRef: any = null;  // ButtonComponent from obsidian
```
Done: [c37d90c](https://github.com/obsidianmd/obsidian-importer/pull/461/commits/c37d90cadf051dc420fe537b7fbad593e53667c2)

Xheldon commented on src/formats/airtable-api.ts:
```diff
@@ -0,0 +1,2148 @@
+		
+		// Get primary field value (processed)
+		// Airtable always uses each table's primary field as note title
+		const primaryFieldValue = extractStringValue(recordFields[primaryFieldName]);
```
Right, done: [8770b9f](https://github.com/obsidianmd/obsidian-importer/pull/461/commits/8770b9f8b33e800aca6e5225fb0d99240553af26)

Xheldon committed:
Airtable API: Recover wrong change of html attr

tgrosinger committed:
Airtable API: Formatting

tgrosinger committed:
Airtable API: Change how loadButton and toggleSelectButton are assigned

Removed unnecessary secondary references.

tgrosinger reviewed (commented):
@Xheldon, I pushed a commit to your branch for the `toggleButtonRef` and `loadButtonRef`. I didn't have any problems with it losing the reference using this method of assignment. Please let me know what you think and if you find a situation where this doesn't work.

tgrosinger commented:
Nice, looks like pretty much all of the feedback has been addressed. I'm going to do a little testing, and hopefully we can get this merged shortly!

tgrosinger committed:
Airtable API: Fix loadButton assignments

The assignment to null is performed before the constructor, but after
the base class contructor. Because the base class contructor is what
calls init, this was causing our assignment to be overwritten to null.

Xheldon commented:
@tgrosinger 

> Can you see the current Group By and preserve that in the import?
Same for the sort? Bases support sorting by multiple properties like Airtable.
In the gallery view, can you set the image property so that it matches what is used in Airtable?
In the gallery view, can you set the properties so that it matches what is shown in Airtable? It may be different from what is shown in the grid/table view.

NO. What you’re referring to are all view‑level details—field order, sorting, grouping, and even the cover settings for Gallery views. The Airtable API does not expose this view‑layer configuration information.

> In the grid view, can you match the column summaries?

I don't quite understand what this means; could you explain it in more detail?

> There's something wrong with Rollups and Linked properties I think....

Unfortunately, Rollup itself doesn't return its formula information via the API, so I can only retrieve the values of the specified columns from its corresponding records and then concatenate them for display. See: https://airtable.com/developers/web/api/field-model#rollup

This is the information about Airtable's Rollup:

<img width="2024" height="1258" alt="CleanShot_2025_12_23_at_21_17_40@2x" src="https://github.com/user-attachments/assets/d2033a1d-1b33-49f3-b62a-9346d8784043" />


Below is the PostMan API response of the table schema, and it indeed does not contain any Formula information:
<img width="2186" height="1810" alt="CleanShot_2025_12_23_at_21_16_45@2x" src="https://github.com/user-attachments/assets/5ab4c337-dbb6-4e5f-8ef1-e73ed670fb3b" />

tgrosinger commented:
> No. What you’re referring to are all view‑level details

Ah okay, that's too bad. Thank you for looking into this.

> I don't quite understand what this means; could you explain it in more detail?

Summaries are displayed in the footer of a table in Obsidian. Here's what they look like:

<img width="860" height="352" alt="image" src="https://github.com/user-attachments/assets/43e45a8e-de05-4163-ad55-48a83e4dedcb" />

[Obsidian summaries docs](https://help.obsidian.md/bases/syntax#Summaries)

In Airtable they are displayed in the header like this:

<img width="1349" height="361" alt="image" src="https://github.com/user-attachments/assets/30c7a7c5-cbb7-42c0-b433-b17c45918bc7" />

> Unfortunately, Rollup itself doesn't return its formula information via the API

Darn. Maybe we should just omit Rollup columns then, since they aren't going to display the correct information. We'll need to note that in the docs.

Xheldon committed:
Import only Rollup property names, ignoring the values.
